infrastructure/smart-contracts/contracts/upgrades/esXai/esXai4.sol

```diff
@@ -0,0 +1,378 @@
+        emit RedemptionStarted(msg.sender, _extRedemptionRequests[msg.sender].length - 1);
+    }
+
+    /**
+     * @dev Function to cancel the redemption process
+     * @param index The index of the redemption request to cancel.
+     */
+    function cancelRedemption(uint256 index) public {
+        require(_redemptionActive, "Redemption is currently inactive");
+        RedemptionRequestExt storage request = _extRedemptionRequests[msg.sender][index];
+        require(request.amount > 0, "Invalid request");
+        require(!request.completed, "Redemption already completed");
+
+        // Mark the redemption request as completed
+        request.completed = true;
+        request.cancelled = true;
+        request.endTime = block.timestamp;
+
+        // Decrement the total pending redemptions
+        _totalPendingRedemptions[msg.sender] -= request.amount;
+
+        emit RedemptionCancelled(msg.sender, index);
+    }
+
+    /**
+     * @dev Function to complete the redemption process
+     * @param index The index of the redemption request to complete.
+     */
+    function completeRedemption(uint256 index) nonReentrant public {
+        require(_redemptionActive, "Redemption is currently inactive");
+        RedemptionRequestExt storage request = _extRedemptionRequests[msg.sender][index];
+        require(request.amount > 0, "Invalid request");
+        require(!request.completed, "Redemption already completed");
+        require(block.timestamp >= request.startTime + request.duration, "Redemption period not yet over");
+
+        // Check if the sender failed KYC
+        bool failedKyc = PoolFactory3(poolFactoryAddress).failedKyc(msg.sender);
+        require(!failedKyc, "KYC failed, cannot redeem");
+
+        // Retrieve the number of licenses owned from the nodeLicense contract
+        uint256 licenseCountOwned = NodeLicense8(nodeLicenseAddress).balanceOf(msg.sender);
+
+        // If the wallet owns more licenses than the maxKeysNonKyc, check if the wallet is KYC approved
+        if(licenseCountOwned > maxKeysNonKyc){
+            Referee9 referee = Referee9(refereeAddress);
+            require(referee.isKycApproved(msg.sender), "You own too many keys, must be KYC approved to claim.");
+        }
+
+        // Calculate the conversion ratio based on the duration
+        uint256 ratio;
+        if (request.duration == 15 days) {
+            ratio = 250;
+        } else if (request.duration == 90 days) {
+            ratio = 625;
+        } else {
+            ratio = 1000;
+        }
+
+        // Calculate the amount of Xai to mint
+        uint256 xaiAmount = request.amount * ratio / 1000;
+
+        // mark the request as completed
+        request.completed = true;
+        request.endTime = block.timestamp;
+
+        // Adding this just in case a conversion was some how missed.
+        // This would allow the user to still redeem the esXai while keeping the state correct.
+        if(request.voucherIssued) {     
```

Comment:
Two questions:
1. Should we keep this as a fail safe?
2. If so, should we also add it in the cancel function?

Reply:
Yeah I don't see why not. That one if is not going to impact the tx costs very much I would hope

Reply:
We'd need it as a require either way, so it has to be checked no matter what.